package-tests.cake

```diff
@@ -0,0 +1,312 @@
+// Representation of a single test to be run against a pre-built package.
+public struct PackageTest
+{
+    public string Description;
+    public string Arguments;
+    public ExpectedResult ExpectedResult;
+
+    public PackageTest(string description, string arguments, ExpectedResult expectedResult)
+    {
+        Description = description;
+        Arguments = arguments;
+        ExpectedResult = expectedResult;
+
+        // Add Common tests here - currently there are none
+        // Derived classes may add more tests
+    }
+}
+
+// Abstract base for all package testers. 
+public abstract class PackageTester
+{
+    protected ICakeContext _context;
+    protected string _packageVersion;
+    protected string _packageDir;
+    protected string _config;
+    protected string _outputDir;
+
+    public PackageTester(ICakeContext context, string packageVersion)
+    {
+        _context = context;
+        _packageVersion = packageVersion;
+        _packageDir = System.IO.Path.GetFullPath(context.Argument("artifact-dir", "package")) + "/";
+        _config = context.Argument("configuration", "Release");
+        _outputDir = System.IO.Path.GetFullPath($"bin/{_config}/");
+
+        PackageTests = new List<PackageTest>();
+    }
+
+    protected abstract string PackageName { get; }
+    protected abstract string PackageInstallDirectory { get; }
+    protected abstract string PackageBinDir { get; }
+
+    protected string PackageUnderTest => _packageDir + PackageName;
+    protected List<PackageTest> PackageTests { get; }
+
+    public void RunTests()
+    {
+        Console.WriteLine("Testing package " + PackageName);
+
+        Console.WriteLine($"Creating Test Directory:\n  {PackageInstallDirectory}");
+        CreatePackageInstallDirectory();
+
+        RunPackageTests();
+    }
+
+    // Default is to just unzip... individual package testers may override
+    protected virtual void CreatePackageInstallDirectory()
+    {
+        _context.CleanDirectory(PackageInstallDirectory);
+        _context.Unzip(PackageUnderTest, PackageInstallDirectory);
+    }
+
+    private void RunPackageTests()
+    {
+        var reporter = new ResultReporter(PackageName);
+
+        foreach (var packageTest in PackageTests)
+        {
+            var resultFile = _outputDir + "TestResult.xml";
+            // Delete result file ahead of time so we don't mistakenly
+            // read a left-over file from another test run. Leave the
+            // file after the run in case we need it to debug a failure.
+            if (_context.FileExists(resultFile))
+                _context.DeleteFile(resultFile);
```

Comment:
Would this also delete the results files from the main nunit-engine unit test run? Or it outputDir something else here?

Reply:
My 'outputDir' is basically bin\Release. I think that's also where our own test results are saved, but we already overwrite them so I didn't figure it was a problem. If we wanted to, we could come back and save __every__ result file to a different place but I'd rather do that as a small, separate PR.

Reply:
Ah ok. Didn't realise we overwrote them already - fair enough, separate issue!